README
Scrypt for Python
==

[![Build Status](https://travis-ci.org/jvarho/pylibscrypt.svg)](https://travis-ci.org/jvarho/pylibscrypt)
[![Coverage Status](https://coveralls.io/repos/github/jvarho/pylibscrypt/badge.svg?branch=master)](https://coveralls.io/github/jvarho/pylibscrypt?branch=master)
[![PyPI version](https://img.shields.io/pypi/v/pylibscrypt.svg)](https://pypi.python.org/pypi/pylibscrypt)

There are a lot of different scrypt modules for Python, but none of them have
everything that I'd like, so here's One More[1].


Features
--
* Uses hashlib.scrypt on Python 3.6 and OpenSSL 1.1.
* Uses system libscrypt[2] as the next choice.
* If neither is available, tries the scrypt Python module[3] or libsodium[4].
* Offers a pure Python scrypt implementation for when there is no C scrypt.
* Not unusably slow, even in pure Python... at least with pypy[5].

With PyPy as the interpreter the Python implementation is around one fifth the
speed of C scrypt. With CPython it is between about 50x and 250x slower.


Requirements
--
* Python 2.7.8+ or 3.4+. Equivalent versions of PyPy should also work.
* If you want speed, you should use one of:
  - libscrypt 1.8+ (older may work)
  - py-scrypt 0.6+ (pip install scrypt)
  - libsodium 1.0+
  - Python 3.6+ with OpenSSL 1.1+


Usage
--

You can install the most recent release from PyPi using:

    pip install pylibscrypt

You most likely want to create MCF hashes and store them somewhere, then check
user-entered passwords against those hashes. For that you only need to use two
functions from the API:

    from pylibscrypt import scrypt_mcf, scrypt_mcf_check
    # Generate an MCF hash with random salt
    mcf = scrypt_mcf(b'Hello World')
    # Test it
    print(scrypt_mcf_check(mcf, b'Hello World'))   # prints True
    print(scrypt_mcf_check(mcf, b'HelloPyWorld'))  # prints False

For full API, you can try help(pylibscrypt) from python after importing.

It is highly recommended that you use a random salt, i.e. don't pass one.


Versioning
--
The package has a version number that can be read from python like so:

    print(pylibscrypt.__version__)

The version number is of the form X.Y.Z, following Semantic Versioning[6].
Releases are tagged vX.Y.Z and release branches bX.Y.x when they differ from
master.


Development
--
Development happens on GitHub[7]. If you find a bug, please open an issue there.

Running pylibscrypt.tests will test all implementations with some quick tests.
Running any implementation directly (e.g. pylibscrypt.pylibsodium) will also
compare to scrypt test vectors from the paper but this is slow for the pure
Python version (pypyscrypt) unless running with pypy.

You can test more comprehensively using the docker test environment. Either
build and run using `make docker-run` or pull the jvarho/pylibscrypt image and
run using `docker run -v ${PWD}:/app jvarho/pylibscrypt`.

Pull requests should be automatically tested and will not be merged if broken.

[1]:https://xkcd.com/927/
[2]:https://github.com/technion/libscrypt
[3]:https://bitbucket.org/mhallin/py-scrypt/src
[4]:https://github.com/jedisct1/libsodium
[5]:http://pypy.org/
[6]:http://semver.org/spec/v2.0.0.html
[7]:https://github.com/jvarho/pylibscrypt



CHANGES
Release History
==

1.8.0
--
2019-03-09

Feature release:
- Support unicode passwords with scrypt_mcf
- Last release to support python 3.3 and <2.7.8


1.7.1
--
2017-12-16

Bugfix release:
- Prevent DepracationWarning from pylibsodium.py


1.7.0
--
2017-11-16

Feature release:
- Improved error messages
- Added coveralls testing
- Improved test coverage using docker
- Deprecated fallbacks for old python and libsodium versions


1.6.1
--
2017-05-08

Bugfix release:
- Fix hashlib.scrypt calls with large values


1.6.0
--
2017-01-04

Feature release:
- Use hashlib.scrypt on Python 3.6 + OpenSSL 1.1
- Travis CI integration
- Improved test coverage
- Code and build cleanups


1.5.3
--
2015-06-19

Bugfix release:
- Fix libsodium incompatibility fallback


1.5.2
--
2015-12-22

Bugfix release:
- Fix library loading on OS/X with packaged apps


1.5.1
--
2015-9-15

Bugfix release:
- Compare MCF hashes in constant time


1.5.0
--
2015-9-7

Feature release:
- Handle dependency failures better
- Improve test coverage
- Misc. cleanups


1.4.1
--
2015-3-27

Bugfix release:
- Fix missing import so Windows/Darwin may actually work


1.4.0
--
2015-3-23

Feature release:
- Improve library loading on Windows and Darwin.
- Obsolete libsodium_salsa implementation.


1.3.0
--
2014-7-8

Feature release:
- Support for libsodium 0.6+ changes.
- Use the low-level scrypt interface with libsodium so all calls are fast.
- Support python3 when running tests.
- Switch to the less verbose ISC license where possible


1.2.1
--
2014-5-17

Bugfix release:
- Fix python3 support.
- Remember to import the libsodium backend added in 1.2.0.


1.2.0
--
2014-5-16

Feature release:
- Support for the $7$ Modular Crypt Format that's been semi-standardized and 
  used by libsodium. (The $s1$ format remains the default.)
- Use libsodium 0.5+ pwhash as another fallback when parameters support it.
- More liberal in reading MCF hashes, so those from other implementations work.
- Fuzz testing of scrypt inputs.


1.1.3
--
2014-5-16

Bugfix release:
- Disallow null in MCF passwords, because that could result in a password that
  libscrypt can't verify.
- Verify MCF passwords with null even when using libscrypt to stay compatible
  with earlier releases.


1.1.2
--
2014-5-12

Bugfix release:
- Fix passwords longer than 256 bytes failing on python2


1.1.1
--
2014-5-12

Bugfix release:
- Fix some r and p values over/underflowing with libscrypt backend.
- Always check that r*p < 2**30.


1.1.0
--
2014-5-8

Feature release:
- On CPython libsodium is used for salsa20/8, for a 4x speedup when C scrypt
  is not available but libsodium is (optional).
- Tests have been converted to unittest. More readable and better coverage.


1.0.3
--
2014-5-5

Bugfix release:
- Guard against some libscrypt bugs found in older versions.
- Fix broken package import when libscrypt isn't available


1.0.2
--
2014-5-3

Packaging fix


1.0.1
--
2014-5-3

Bugfix release:
- Enforced parameter requirements that libscrypt has for MCF form hashes.
  Initial release could accept parameters for scrypt_mcf that scrypt_mcf_check
  didn't support when using libscrypt backend.
- Fixed documentation and tests.


1.0.0
--
2014-5-2

Initial release



pylibscrypt/fuzz.py
# Copyright (c) 2014-2015, Jan Varho
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

"""Fuzzes scrypt function input, comparing two implementations"""

import itertools
import random
from random import randrange as rr
import unittest


class Skip(Exception):
    pass


class Fuzzer(object):
    """Fuzzes function input"""
    def __init__(self, f, args, g=None, pass_good=None, pass_bad=None):
        self.f = f
        self.g = g
        self.args = args
        self.pass_good = pass_good
        self.pass_bad = pass_bad

    @staticmethod
    def get_random_int():
        return int((1<<rr(66)) * 1.3)

    @staticmethod
    def get_random_bytes(lrange=None, skip=None):
        if lrange is None:
            v = bytearray(rr(2**rr(10)))
        else:
            v = bytearray(rr(*lrange))
        for i in range(len(v)):
            v[i] = rr(256)
            while v[i] == skip:
                v[i] = rr(256)
        return bytes(v)

    def get_good_args(self):
        kwargs = {}
        for a in self.args:
            assert isinstance(a, dict)
            if 'opt' in a and a['opt'] and random.randrange(2):
                continue
            if 'val' in a:
                kwargs[a['name']] = a['val']
            elif 'vals' in a:
                kwargs[a['name']] = random.choice(a['vals'])
            elif 'valf' in a:
                kwargs[a['name']] = a['valf']()
            elif 'type' in a and a['type'] == 'int':
                kwargs[a['name']] = self.get_random_int()
            elif 'type' in a and a['type'] == 'bytes':
                kwargs[a['name']] = self.get_random_bytes()
            else:
                raise ValueError
            if 'none' in a and not random.randrange(10):
                kwargs[a['name']] = None
            if 'skip' in a and a['skip'](kwargs[a['name']]):
                if 'opt' in a and a['opt']:
                    del kwargs[a['name']]
                else:
                    raise Skip
        return kwargs

    def get_bad_args(self, kwargs=None):
        kwargs = kwargs or self.get_good_args()
        a = random.choice(self.args)
        if not 'opt' in a:
            if not random.randrange(10):
                del kwargs[a['name']]
                return kwargs
        if not 'type' in a:
            return self.get_bad_args(kwargs)

        if not random.randrange(10):
            wrongtype = [
                self.get_random_int(), self.get_random_bytes(), None,
                1.1*self.get_random_int(), 1.0*self.get_random_int()
            ]
            if a['type'] == 'int':
                del wrongtype[0]
            elif a['type'] == 'bytes':
                del wrongtype[1]
            v = random.choice(wrongtype)
            try:
                if 'valf' in a:
                    if a['valf'](v):
                        return self.get_bad_args(kwargs)
                if 'skip' in a and a['skip'](v):
                    return self.get_bad_args(kwargs)
            except TypeError:
                pass # Surely bad enough
            kwargs[a['name']] = v
            return kwargs

        if a['type'] == 'int':
            v = self.get_random_int()
            if 'valf' in a:
                if a['valf'](v):
                    return self.get_bad_args(kwargs)
            if 'skip' in a and a['skip'](v):
                return self.get_bad_args(kwargs)
            kwargs[a['name']] = v
            return kwargs

        if a['type'] == 'bytes' and 'valf' in a:
            v = self.get_random_bytes()
            if not a['valf'](v):
                kwargs[a['name']] = v
                return kwargs

        return self.get_bad_args(kwargs)

    def fuzz_good_run(self, tc):
        try:
            kwargs = self.get_good_args()
            r1 = self.f(**kwargs)
            r2 = self.g(**kwargs) if self.g is not None else None
            if self.g is not None:
                r2 = self.g(**kwargs)
        except Skip:
            tc.skipTest('slow')
        except Exception as e:
            assert False, ('unexpected exception', kwargs, e)

        try:
            if self.pass_good:
                tc.assertTrue(self.pass_good(r1, r2, kwargs),
                              msg=('unexpected output', r1, r2, kwargs))
            else:
                if self.g is not None:
                    assert r1 == r2, ('f and g mismatch', kwargs, r1, r2)
                tc.assertTrue(r1)
        except Exception as e:
            print ('unexpected exception', kwargs, r1, r2, e)
            raise

    def fuzz_bad(self, f=None, kwargs=None):
        f = f or self.f
        kwargs = kwargs or self.get_bad_args()
        return f(**kwargs)

    def fuzz_bad_run(self, tc):
        try:
            kwargs = self.get_bad_args()
        except Skip:
            tc.skipTest('slow')
        for f in ((self.f,) if not self.g else (self.f, self.g)):
            try:
                r = self.fuzz_bad(f, kwargs)
                assert False, ('no exception', kwargs, r)
            except Skip:
                tc.skipTest('slow')
            except AssertionError:
                raise
            except Exception:
                pass

    def testcase_good(self, tests=1, name='FuzzTestGood'):
        testfs = {}
        for i in range(tests):
            testfs['test_fuzz_good_%d' % i] = lambda s: self.fuzz_good_run(s)
        t = type(name, (unittest.TestCase,), testfs)
        return t

    def testcase_bad(self, tests=1, name='FuzzTestBad'):
        testfs = {}
        for i in range(tests):
            testfs['test_fuzz_bad_%d' % i] = lambda s: self.fuzz_bad_run(s)
        t = type(name, (unittest.TestCase,), testfs)
        return t

    def generate_tests(self, suite, count, name):
        loader = unittest.defaultTestLoader
        suite.addTest(
            loader.loadTestsFromTestCase(
                self.testcase_good(count, name + 'Good')
            )
        )
        suite.addTest(
            loader.loadTestsFromTestCase(
                self.testcase_bad(count, name + 'Bad')
            )
        )



if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description='Fuzz testing')
    parser.add_argument('-c', '--count', type=int, default=100)
    parser.add_argument('-f', '--failfast', action='store_true')
    clargs = parser.parse_args()

    modules = []
    try:
        from . import pylibscrypt
        modules.append((pylibscrypt, 'pylibscrypt'))
    except ImportError:
        pass

    try:
        from . import pyscrypt
        modules.append((pyscrypt, 'pyscrypt'))
    except ImportError:
        pass

    try:
        from . import pylibsodium_salsa
        modules.append((pylibsodium_salsa, 'pylibsodium_salsa'))
    except ImportError:
        pass

    try:
        from . import pylibsodium
        modules.append((pylibsodium, 'pylibsodium'))
    except ImportError:
        pass

    try:
        from . import pypyscrypt_inline as pypyscrypt
        modules.append((pypyscrypt, 'pypyscrypt'))
    except ImportError:
        pass

    scrypt_args = (
        {'name':'password', 'type':'bytes'},
        {'name':'salt', 'type':'bytes'},
        {
            'name':'N', 'type':'int', 'opt':False,
            'valf':(lambda N=None: 2**rr(1,6) if N is None else
                    1 < N < 2**64 and not (N & (N - 1))),
            'skip':(lambda N: (N & (N - 1)) == 0 and N > 32 and N < 2**64)
        },
        {
            'name':'r', 'type':'int', 'opt':True,
            'valf':(lambda r=None: rr(1, 16) if r is None else 0<r<2**30),
            'skip':(lambda r: r > 16 and r < 2**30)
        },
        {
            'name':'p', 'type':'int', 'opt':True,
            'valf':(lambda p=None: rr(1, 16) if p is None else 0<p<2**30),
            'skip':(lambda p: p > 16 and p < 2**30)
        },
        {
            'name':'olen', 'type':'int', 'opt':True,
            'valf':(lambda l=None: rr(1, 1000) if l is None else l >= 0),
            'skip':(lambda l: l < 0 or l > 1024)
        },
    )

    scrypt_mcf_args = (
        {
            'name':'password', 'type':'bytes',
            'valf':(lambda p=None: Fuzzer.get_random_bytes(skip=0) if p is None
                    else not b'\0' in p)
        },
        {
            'name':'salt', 'type':'bytes', 'opt':False, 'none':True,
            'valf':(lambda s=None: Fuzzer.get_random_bytes((1,17)) if s is None
                    else 1 <= len(s) <= 16)
        },
        {
            'name':'N', 'type':'int', 'opt':False,
            'valf':(lambda N=None: 2**rr(1,6) if N is None else
                    1 < N < 2**64 and not (N & (N - 1))),
            'skip':(lambda N: (N & (N - 1)) == 0 and N > 32 and N < 2**64)
        },
        {
            'name':'r', 'type':'int', 'opt':True,
            'valf':(lambda r=None: rr(1, 16) if r is None else 0<r<2**30),
            'skip':(lambda r: r > 16 and r < 2**30)
        },
        {
            'name':'p', 'type':'int', 'opt':True,
            'valf':(lambda p=None: rr(1, 16) if p is None else 0<p<2**30),
            'skip':(lambda p: p > 16 and p < 2**30)
        },
        {
            'name':'prefix', 'type':'bytes', 'opt':True,
            'vals':(b'$s1$', b'$7$'),
            'skip':(lambda p: p is None)
        }
    )

    random.shuffle(modules)
    suite = unittest.TestSuite()
    loader = unittest.defaultTestLoader
    for m1, m2 in itertools.permutations(modules, 2):
        Fuzzer(
            m1[0].scrypt, scrypt_args, m2[0].scrypt,
            pass_good=lambda r1, r2, a: (
                isinstance(r1, bytes) and
                (r2 is None or r1 == r2) and
                (len(r1) == 64 if 'olen' not in a else len(r1) == a['olen'])
            )
        ).generate_tests(suite, clargs.count, m1[1])
        Fuzzer(
            m1[0].scrypt_mcf, scrypt_mcf_args, m2[0].scrypt_mcf,
            pass_good=lambda r1, r2, a: (
                m2[0].scrypt_mcf_check(r1, a['password']) and
                (r2 is None or m1[0].scrypt_mcf_check(r2, a['password'])) and
                (r2 is None or 'salt' not in a or a['salt'] is None or r1 == r2)
            )
        ).generate_tests(suite, clargs.count, m1[1])
    unittest.TextTestRunner(failfast=clargs.failfast).run(suite)

